File: app/src/test/java/igrek/songbook/service/songtree/SongTreeFilterTest.java
package igrek.songbook.service.songtree;

import org.junit.Test;
import org.mockito.Mockito;

import static org.assertj.core.api.Java6Assertions.assertThat;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class SongTreeFilterTest {
	
	@Test
	public void test_matchesNameFilter() {
		
		SongTreeItem songItem = mock(SongTreeItem.class, Mockito.RETURNS_DEEP_STUBS);
		
		when(songItem.getSong().getCategory().getDisplayName()).thenReturn("Budka suflera");
		when(songItem.getSong().getTitle()).thenReturn("Jolka jolka ążśźęćół ĄĄŻŚŹĘĆ Żółć Łódź");
		// test mockito
		assertThat(songItem.getSong().getCategory().getDisplayName()).isEqualTo("Budka suflera");
		
		assertThat(new SongTreeFilter("Budka").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("budka").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("uFL udK").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("jolka suflera").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("dupka").matchesNameFilter(songItem)).isFalse();
		assertThat(new SongTreeFilter("dupka suflera").matchesNameFilter(songItem)).isFalse();
		// polish letters
		assertThat(new SongTreeFilter("żółć łÓDŹ").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("zolc").matchesNameFilter(songItem)).isTrue();
		assertThat(new SongTreeFilter("azszecol aazszec lodz zolc").matchesNameFilter(songItem)).isTrue();
		
	}
}

File: app/src/test/java/igrek/songbook/service/songtree/SongTreeFilterTest.kt
package igrek.songbook.service.songtree

import org.assertj.core.api.Assertions.assertThat
import igrek.songbook.persistence.general.model.Category
import igrek.songbook.persistence.general.model.CategoryType
import igrek.songbook.persistence.general.model.Song
import igrek.songbook.persistence.general.model.SongStatus
import igrek.songbook.songselection.search.SongTreeFilter
import igrek.songbook.songselection.tree.SongTreeItem
import org.junit.Test

class SongTreeFilterTest {

    @Test
    fun test_matchesNameFilter() {
        val songItem = SongTreeItem.song(Song(
                id = 1,
                title = "Jolka jolka ążśźęćół ĄĄŻŚŹĘĆ Żółć Łódź",
                categories = mutableListOf(
                        Category(1, type = CategoryType.ARTIST, name = "Budka suflera")
                ),
                status = SongStatus.PUBLISHED
        ))

        assertThat(songItem.song!!.categories[0].name).isEqualTo("Budka suflera")

        assertThat(songItem.song!!.displayName())
                .isEqualTo("Jolka jolka ążśźęćół ĄĄŻŚŹĘĆ Żółć Łódź - Budka suflera")

        assertThat(SongTreeFilter("Budka").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("budka").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("uFL udK").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("jolka suflera").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("dupka").songMatchesNameFilter(songItem)).isFalse()
        assertThat(SongTreeFilter("dupka suflera").songMatchesNameFilter(songItem)).isFalse()
        // polish letters
        assertThat(SongTreeFilter("żółć łÓDŹ").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("zolc").songMatchesNameFilter(songItem)).isTrue()
        assertThat(SongTreeFilter("azszecol aazszec lodz zolc").songMatchesNameFilter(songItem))
                .isTrue()
    }

    @Test
    fun test_filteringWithQuotes() {
        val songItem = SongTreeItem.song(Song(
                id = 1,
                title = "he's dupa",
                categories = mutableListOf(
                        Category(1, type = CategoryType.ARTIST, name = "Budka suflera")
                ),
                status = SongStatus.PUBLISHED
        ))

        assertThat(SongTreeFilter("d'upa hes").songMatchesNameFilter(songItem)).isTrue()
    }
}
